<a href="https://colab.research.google.com/github/ajithbobby/Iris_Flower_Classification/blob/main/AJITH_BOBBY.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Here we have imported some libraries:

In [ ]:
import pandas as pd
import numpy as np

The data we will use is the iris dataset, a classical dataset in machine learning and statistics.It is included in the scikit-learn in the datatsets module.

In [ ]:
from sklearn.datasets import load_iris
iris_dataset= load_iris()

The iris object that is returned by load_iris is a Bunch object, which is very similar to a dictionary. It contains keys and values.

In [ ]:
print("Keys of iris_dataset: \n{}".format(iris_dataset.keys()))

Keys of iris_dataset: 
dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names', 'filename'])


The value of the key DESCR is a short description of the dataset.

In [ ]:
print(iris_dataset['DESCR'][:193] +"\n...")

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, pre
...


The value of the key target_names is an array of strings,containing the species of flower that we want to predict,i.e,'setosa','versicolor' and 'virginica'.

In [ ]:
print("Target names:{}".format(iris_dataset['target_names']))

Target names:['setosa' 'versicolor' 'virginica']


The value of feature_names is a list of strings,giving the description of each feature. It includes 'sepal length','sepal width','petal length' and 'petal width' all in centimeters.

In [ ]:
print("Feature names: \n{}".format(iris_dataset['feature_names']))

Feature names: 
['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


The data that we are going to use is itself contained in the target and data fields.The data contains numeric measurements of sepal length,sepal width,petal length and petal width in a NumPy array.

In [ ]:
print("Type of data:{}".format(type(iris_dataset['data'])))

Type of data:<class 'numpy.ndarray'>


The rows in this data array correspond to the flowers, while the columns represent the four measurements that were taken for each flower.

In [ ]:
print("Shape of data:{}".format(iris_dataset['data'].shape))

Shape of data:(150, 4)


We see that the array contains measurements for 150 different flowers.The individual items are called samples/data points in machine learning, and their properties are called features.So we have 150 data points and 4 features.

In [ ]:
print("First five columns of data:\n{}".format(iris_dataset['data'][:5]))

First five columns of data:
[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]]


From this data, we can see that all of the first five flowers have a petal width of 0.2cm and that the first flower has the longest sepal, at 5.1cm.

In [ ]:
print("Type of target:{}".format(type(iris_dataset['target'])))

Type of target:<class 'numpy.ndarray'>


target is a one-dimensional array, and on seeing the shape we can see that it contains one entry per flower.

In [ ]:
print("Shape of target:{}".format(iris_dataset['target'].shape))

Shape of target:(150,)


On seeing the target key and exploring the values we can easily see that the species are encoded as integers from 0 to 2.

In [ ]:
print("Target:\n{}".format(iris_dataset['target']))

Target:
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


The meanings of the numbers are given by the iris ['target_names'] array:0 means setosa,1 means versicolor and 2 means virginica.

We want to build a machine learning model from this data that can predict the series of iris for a new set of measurements.
In skicit-learn data is usually denoted with a capital X, while labels are denoted by a lowercase y.This is inspired by the standard formulation (x)=y in mathematics,where x is the input of a function and y is the output.Following more conventions from mathematics,we use a capital X because data is a two-dimensional array (a matrix) and a lowercase y because  the target is a one-dimensional array (a vector).

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
                   iris_dataset['data'],iris_dataset['target'], random_state=0)

Before making the split, the train_test_split function shuffles the data set using a pseudorandom number generator. To make sure that we will get the same output if we run the same function several times, we provide the pseudorandom number generator with a fixed seed using the random_state parameter.
The output of the train_test_split function is X_train, X_test, y_train, y_test, which are all Numpy arrays.X_train contains 75% of the rows of the dataset, and X_test contains the remaining 25%.

In [ ]:
print("X_train shape: {}".format(X_train.shape))
print("y_train shape: {}".format(y_train.shape))
print("X_test shape: {}".format(X_test.shape))
print("y_test shape: {}".format(y_test.shape))

X_train shape: (112, 4)
y_train shape: (112,)
X_test shape: (38, 4)
y_test shape: (38,)


Next, we are going to import the machine learning model.

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
knn= KNeighborsClassifier(n_neighbors=1)

The knn model encapsulates the algorithm  that will be used to build the model from the training data, as well as the algorithm to make predictions on new data points. It wil also hold the information that the algorithm has extracted from the training data. In the case of KNeighborsClassifier, it will just store the training set. To build the model on the training set, we call the fit method of the knn object, which takes as arguments the NumPy array X_train containing the training data and the NumPy array y_train of the corresponding training labels: 

In [ ]:
knn.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=1, p=2,
                     weights='uniform')

We can now make predictions using this model on new data for which we might not know the correct labels. Imagine we found an iris in the wild with a sepal length of 5 cm, a sepal width of 2.9cm, a petal length of 1cm and a petal width of 0.2cm.
We can put this data into a NumPy array, again by calculating the shape,i.e, the number of samples(1) multiplied by the number of features(4):

In [ ]:
X_new = np.array([[5, 2.9 ,1, 0.2]])
print("X_new.shape: {}".format(X_new.shape))

X_new.shape: (1, 4)


We made the measurements of this single flower into a row in a two-dimensional NumPy array, as scikit-learn always expects two-dimensional arrays for the data.
To make a prediction, we call the predict method of the knn object:

In [ ]:
prediction= knn.predict(X_new)

print("Prediction: {}".format(prediction))
print("Predicted target name: {}".format(
iris_dataset['target_names'][prediction]))

Prediction: [0]
Predicted target name: ['setosa']


This is where the test set that we created earlier comes in. This data was not used to build the model, but we do not know what the correct species is for each iris in the test set.
Therefore, we can make a prediction for each iris in the test data and compare it against its label( the known species). We can measure how well the model works by computing the accuracy ,which is the fraction of flowers for which the right species was predicted:

In [ ]:
y_pred = knn.predict(X_test)
print("Test set prediction:\n {}".format(y_pred))

Test set prediction:
 [2 1 0 2 0 2 0 1 1 1 2 1 1 1 1 0 1 1 0 0 2 1 0 0 2 0 0 1 1 0 2 1 0 2 2 1 0
 2]


In [ ]:
print("Test set score: {}".format(np.mean(y_pred == y_test)))

Test set score: 0.9736842105263158


From this, we found out that accuracy of our model which we have created is 97.368%.